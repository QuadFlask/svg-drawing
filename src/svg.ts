import { download } from './utils/download'
const roundUp = (num: number) => Number(num.toFixed(2))

export class Point {
  public x: number
  public y: number
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  public toVector(): Vector {
    const value = Math.sqrt(Math.pow(this.x, 2.0) + Math.pow(this.y, 2.0))
    const angle = Math.atan2(this.y, this.x)
    return new Vector(value, angle)
  }

  public scale(r: number): Point {
    return new Point(this.x * r, this.y * r)
  }

  public add(p: Point): Point {
    return new Point(this.x + p.x, this.y + p.y)
  }

  public sub(p: Point): Point {
    return new Point(this.x - p.x, this.y - p.y)
  }
  public commandMove() {
    return `M ${this.x} ${this.y}`
  }

  public commandLine() {
    return `L ${this.x} ${this.y}`
  }

  public commandCirculer(cl: Point, cr: Point): string {
    return `C ${cl.x} ${cl.y} ${cr.x} ${cr.y} ${this.x} ${this.y}`
  }
}

export enum CommandType {
  MOVE,
  LINE,
  CURVE
}

export class Command {
  public point: Point
  public type: CommandType
  public cl?: Point
  public cr?: Point
  constructor(type: CommandType, point: Point) {
    this.point = point
    this.type = type
  }

  public toString(): string {
    switch (this.type) {
      case CommandType.MOVE:
        return this.point.commandMove()
      case CommandType.LINE:
        return this.point.commandLine()
      case CommandType.CURVE:
        if (!this.cl || !this.cr) return this.point.commandLine()
        return this.point.commandCirculer(this.cl, this.cr)
    }
  }
}
export class Vector {
  public angle: number
  public value: number
  constructor(v: number, a: number) {
    this.value = v
    this.angle = a
  }

  public toPoint(): Point {
    const x = Math.cos(this.angle) * this.value
    const y = Math.sin(this.angle) * this.value
    return new Point(roundUp(x), roundUp(y))
  }

  public scale(r: number): Vector {
    return new Vector(this.value * r, this.angle)
  }
}

export interface PathOption {
  close?: boolean
  circuler?: boolean
  strokeWidth?: number
  stroke?: string
  fill?: string
}
export class Path {
  public close: boolean
  public circuler: boolean
  public strokeWidth: number
  public stroke: string
  public fill: string
  public smoothRatio: number
  public points: Point[]
  public commands: Command[]

  constructor({ close, circuler, strokeWidth, stroke, fill }: PathOption = {}) {
    this.close = close ?? false
    this.circuler = circuler ?? true
    this.strokeWidth = strokeWidth ?? 1
    this.stroke = stroke ?? '#000'
    this.fill = fill ?? 'none'
    this.smoothRatio = 0.2
    this.points = []
    this.commands = []
  }

  public scale(r: number) {
    const update = this.points.map((p: Point) => p.scale(r))
    this.points = update
    this.strokeWidth *= r
  }

  public addPoint(p: Point) {
    this.points.push(p)
  }

  public addCommand(c: Command) {
    this.commands.push(c)
  }
  public createCommand(): string {
    this.formatCommand()
    if (this.commands.length === 0) return ''

    let d = this.commands
      .map((com: Command, _i: number) => com.toString())
      .join(' ')
    if (this.close) {
      d += ` Z`
    }
    return d
  }

  public formatCommand() {
    const isCirculer = this.circuler && this.points.length > 2
    const points = this.close
      ? [...this.points, this.points[0]]
      : [...this.points]
    this.commands = isCirculer
      ? this._createCurveCommand(points)
      : this._createLineCommand(points)
  }

  private _createControlPoint(prev: Point, start: Point, next: Point): Point {
    const contolVectorPoint = next
      .sub(prev)
      .toVector()
      .scale(this.smoothRatio)
      .toPoint()
    return start.add(contolVectorPoint)
  }
  private _createCurveCommand(pts: Point[]): Command[] {
    let update = []
    const endIndex = pts.length - 1
    update.push(new Command(CommandType.MOVE, this.points[0]))
    for (let i = 1; i < pts.length; i += 1) {
      const p1 = i === 1 ? pts[0] : pts[i - 2]
      const p2 = pts[i - 1]
      const p3 = pts[i]
      // TODO: Refactor
      const p4 = this.close
        ? i === endIndex
          ? pts[1]
          : i === endIndex - 1
          ? pts[i]
          : pts[i + 1]
        : i === endIndex
        ? pts[i]
        : pts[i + 1]
      const curveCommand = new Command(CommandType.CURVE, pts[i])
      curveCommand.cl = this._createControlPoint(p1, p2, p3)
      curveCommand.cr = this._createControlPoint(p4, p3, p2)
      update.push(curveCommand)
    }
    return update
  }

  private _createLineCommand(pts: Point[]): Command[] {
    let update = []
    update.push(new Command(CommandType.MOVE, pts[0]))
    for (let i = 1; i < pts.length; i += 1) {
      update.push(new Command(CommandType.LINE, pts[i]))
    }
    return update
  }

  public toElement(): HTMLElement {
    const path = document.createElement('path')
    path.setAttribute('stroke-width', String(this.strokeWidth))
    path.setAttribute('stroke', this.stroke)
    path.setAttribute('fill', this.fill)
    path.setAttribute('stroke-linejoin', this.circuler ? 'round' : 'mitter')
    path.setAttribute('stroke-linecap', this.circuler ? 'round' : 'square')
    path.setAttribute('d', this.createCommand())
    return path
  }

  public clone(): Path {
    const path = new Path({
      close: this.close,
      circuler: this.circuler,
      strokeWidth: this.strokeWidth,
      stroke: this.stroke,
      fill: this.fill
    })
    console.log(path)
    this.points.map(p => {
      path.addPoint(p)
    })
    this.commands.map(c => {
      path.addCommand(c)
    })
    return path
  }
}

export interface SvgOption {
  width: number
  height: number
  background?: string
}
export class Svg {
  public background: string
  private _width: number
  private _height: number
  private _paths: Path[]
  constructor({ width, height, background }: SvgOption) {
    this._paths = []
    this._width = width
    this._height = height
    this.background = background ?? '#fff'
    // this.scalePath = this.scalePath.bind(this)
    // this.toElement = this.toElement.bind(this)
    // this.toBase64 = this.toBase64.bind(this)
    // this.resizeElement = this.resizeElement.bind(this)
  }

  public resizeElement(width: number, height: number) {
    this._width = width
    this._height = height
    // TODO: Resizing improve
    if (this._width === width) return
    this.scalePath(width / this._width)
  }

  public scalePath(r: number) {
    for (let i = 0; i < this._paths.length; i += 1) {
      this._paths[i].scale(r)
    }
  }

  public addPath(pa: Path) {
    this._paths.push(pa)
  }

  public undoPath(): Path | undefined {
    return this._paths.pop()
  }

  public replacePath(paths: Path[]) {
    this._paths = paths
  }

  public addPoint(po: Point) {
    if (this._paths.length === 0) return
    const updateIndex = this._paths.length - 1
    this._paths[updateIndex].addPoint(po)
  }

  public updatePath(pa: Path, i?: number) {
    const updateIndex = i || this._paths.length - 1
    if (updateIndex < 0) this._paths.push(pa)
    this._paths[updateIndex] = pa
  }

  public clearPath() {
    this._paths = []
  }

  public get paths(): Path[] {
    return this._paths
  }

  public toElement(): SVGSVGElement {
    const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg')
    svg.setAttribute('width', String(this._width))
    svg.setAttribute('height', String(this._height))
    for (let i = 0; i < this._paths.length; i += 1) {
      svg.appendChild(this._paths[i].toElement())
    }
    return svg
  }

  public toBase64(): string {
    const data = `<svg width="${this._width}" height="${
      this._height
    }" version="1.1" xmlns="http://www.w3.org/2000/svg">${
      this.toElement().innerHTML
    }</svg>`
    return `data:image/svg+xml;base64,${btoa(data)}`
  }

  // TODO: Add filename config
  public download(
    ext: 'svg' | 'png' | 'jpg' = 'svg',
    cb: typeof download = download
  ) {
    if (ext === 'svg') {
      cb({
        data: this.toBase64(),
        extension: 'svg'
      })
    }

    const img: any = new Image()
    const renderCanvas = () => {
      const canvas = document.createElement('canvas')
      canvas.setAttribute('width', String(this._width))
      canvas.setAttribute('height', String(this._height))
      const ctx = canvas.getContext('2d')
      if (!ctx) return
      ctx.fillStyle = this.background
      ctx.fillRect(0, 0, this._width, this._height)
      ctx.drawImage(img, 0, 0)
      if (ext === 'png') {
        cb({ data: canvas.toDataURL('image/png'), extension: 'png' })
      } else {
        cb({ data: canvas.toDataURL('image/jpeg'), extension: 'jpg' })
      }
    }
    img.addEventListener('load', renderCanvas, false)
    img.src = this.toBase64()
  }
}
