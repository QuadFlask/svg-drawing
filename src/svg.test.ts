import { Point, Vector, Path, Svg, Command, CommandType } from './svg'

describe('svg', () => {
  describe('Point', () => {
    it('add', () => {
      const po = new Point(1.0, 1.0).add(new Point(2.0, 2.0))
      expect(po.x).toBe(3.0)
      expect(po.y).toBe(3.0)
    })

    it('sub', () => {
      const po = new Point(3.0, 3.0).sub(new Point(1.0, 1.0))
      expect(po.x).toBe(2.0)
      expect(po.y).toBe(2.0)
    })

    it('scale', () => {
      const po = new Point(1.0, 1.0).scale(3)
      expect(po.x).toBe(3.0)
      expect(po.y).toBe(3.0)
    })

    it('toVector', () => {
      const vec = new Point(1.0, 1.0).toVector()
      expect(vec.value).toBe(1.4142135623730951)
      expect(vec.angle).toBe(0.7853981633974483)
    })
  })
  describe('Vector', () => {
    it('toPoint', () => {
      const po = new Vector(1.4142135, 0.7853982).toPoint()
      expect(po.x).toBe(1.0)
      expect(po.y).toBe(1.0)
    })
    it('scale', () => {
      const vec = new Vector(1.0, 0.5).scale(0.3)
      expect(vec.value).toBe(0.3)
      expect(vec.angle).toBe(0.5)
    })
  })
  describe('Command', () => {
    it('MOVE', () => {
      const commands = new Command(CommandType.MOVE, new Point(0, 0))
      expect(commands.toString()).toBe('M 0 0')
    })
    it('LINE', () => {
      const commands = new Command(CommandType.LINE, new Point(1, 1))
      expect(commands.toString()).toBe('L 1 1')
    })
    it('CURVE', () => {
      const commands = new Command(CommandType.CURVE, new Point(1, 1))
      expect(commands.toString()).toBe('L 1 1')
      commands.cl = new Point(0.25, 0.25)
      commands.cr = new Point(0.75, 0.25)
      expect(commands.toString()).toBe('C 0.25 0.25 0.75 0.25 1 1')
    })
  })
  describe('Path', () => {
    it('addPoint', () => {
      const path = new Path().addPoint(new Point(1, 1))
      expect(path.points.length).toBe(1)
      expect(path.points[0].x).toBe(1)
      expect(path.points[0].y).toBe(1)
    })
    it('addCommand', () => {
      const path = new Path().addCommand(
        new Command(CommandType.LINE, new Point(1, 1))
      )
      expect(path.commands.length).toBe(1)
      expect(path.commands[0].type).toBe(CommandType.LINE)
      expect(path.commands[0].point.x).toBe(1)
      expect(path.commands[0].point.y).toBe(1)
    })
    it('scale', () => {
      const path = new Path({ strokeWidth: 1 })
        .addPoint(new Point(1, 1))
        .scale(2)
      expect(path.strokeWidth).toBe(2)
      expect(path.points[0].x).toBe(2)
      expect(path.points[0].y).toBe(2)
    })
    it('clone', () => {
      const path = new Path({ strokeWidth: 1 }).addPoint(new Point(1, 1))
      const clone = path.clone().addPoint(new Point(2, 2))
      clone.points[0].x = 3
      clone.formatCommand()
      expect(path.points.length).toBe(1)
      expect(path.commands.length).toBe(0)
      expect(path.points[0].x).toBe(1)
      expect(clone.points.length).toBe(2)
      expect(clone.commands.length).toBe(2)
      expect(clone.points[0].x).toBe(3)
    })
    describe('toJson and toElement', () => {
      const path = new Path({ circuler: true, close: false })
        .addPoint(new Point(0, 0))
        .addPoint(new Point(1, 1))
        .addPoint(new Point(2, 1))
        .addPoint(new Point(3, 0))
        .formatCommand()
      it('toJson', () => {
        expect(path.toJson()).toMatchSnapshot()
      })
      it('toElement', () => {
        expect(path.toElement()).toMatchSnapshot()
      })
    })
    describe('formatCommand and getCommandString', () => {
      describe('Line', () => {
        const path = new Path({ circuler: false })
          .addPoint(new Point(0, 0))
          .addPoint(new Point(1, 1))
          .addPoint(new Point(-1, -1))
        it('Normal', () => {
          path.close = false
          expect(path.formatCommand().commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
        it('Close', () => {
          path.close = true
          expect(path.formatCommand().commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
      })
      describe('Circuler', () => {
        const path = new Path({ circuler: true })
          .addPoint(new Point(0, 0))
          .addPoint(new Point(1, 1))
          .addPoint(new Point(2, 1))
          .addPoint(new Point(3, 0))
        it('Normal', () => {
          path.close = false
          expect(path.formatCommand().commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
        it('Close', () => {
          path.close = true
          expect(path.formatCommand().commands).toMatchSnapshot()
          expect(path.getCommandString()).toMatchSnapshot()
        })
      })
    })
  })
  describe('Renderer', () => {
    let svg: Svg
    beforeEach(() => {
      svg = new Svg({ width: 500, height: 500 })
        .addPath(
          new Path({ circuler: true, close: false })
            .addPoint(new Point(0, 0))
            .addPoint(new Point(1, 1))
            .addPoint(new Point(2, 1))
            .addPoint(new Point(3, 0))
            .formatCommand()
        )
        .addPath(
          new Path({ circuler: false, close: true })
            .addPoint(new Point(4, 4))
            .addPoint(new Point(9, 4))
            .addPoint(new Point(9, 8))
            .addPoint(new Point(3, 0))
            .formatCommand()
        )
    })
    // TODO: Fix width, height
    it('toElement', () => {
      expect(svg.toElement()).toMatchSnapshot()
    })
    it('toJson', () => {
      expect(svg.toJson()).toMatchSnapshot()
    })
    // TODO: replace image snapshot
    it('toBase64', () => {
      expect(svg.toBase64()).toMatchSnapshot()
    })
  })
})
