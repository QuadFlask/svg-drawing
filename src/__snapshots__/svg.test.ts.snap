// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svg Path toElement 1`] = `
<path
  d="M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0"
  fill="none"
  stroke="#000"
  stroke-linecap="round"
  stroke-linejoin="round"
  stroke-width="1"
/>
`;

exports[`svg Renderer download svg 1`] = `
Object {
  "data": "data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCIgdmVyc2lvbj0iMS4xIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxwYXRoIHN0cm9rZS13aWR0aD0iMSIgc3Ryb2tlPSIjMDAwIiBmaWxsPSJub25lIiBzdHJva2UtbGluZWpvaW49InJvdW5kIiBzdHJva2UtbGluZWNhcD0icm91bmQiIGQ9Ik0gMCAwIEMgMC4yIDAuMiAwLjYgMC44IDEgMSBDIDEuNCAxLjIgMS42IDEuMiAyIDEgQyAyLjQgMC44IDIuOCAwLjIgMyAwIj48L3BhdGg+PC9zdmc+",
  "extension": "svg",
}
`;

exports[`svg Renderer toBase64 1`] = `"data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAwIiBoZWlnaHQ9IjUwMCIgdmVyc2lvbj0iMS4xIiB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciPjxwYXRoIHN0cm9rZS13aWR0aD0iMSIgc3Ryb2tlPSIjMDAwIiBmaWxsPSJub25lIiBzdHJva2UtbGluZWpvaW49InJvdW5kIiBzdHJva2UtbGluZWNhcD0icm91bmQiIGQ9Ik0gMCAwIEMgMC4yIDAuMiAwLjYgMC44IDEgMSBDIDEuNCAxLjIgMS42IDEuMiAyIDEgQyAyLjQgMC44IDIuOCAwLjIgMyAwIj48L3BhdGg+PC9zdmc+"`;

exports[`svg Renderer toElement 1`] = `
<svg
  height="500"
  width="500"
>
  <path
    d="M 0 0 C 0.2 0.2 0.6 0.8 1 1 C 1.4 1.2 1.6 1.2 2 1 C 2.4 0.8 2.8 0.2 3 0"
    fill="none"
    stroke="#000"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="1"
  />
</svg>
`;
